### PR TITLE
feat: add translations for emails and type error fixes overall

### PR DESCRIPTION
**Pull request by mihaic195**
- I still need to translate the dates displayed in emails (day and month), but it's already hard to follow the PR. We can do it in a follow-up PR.

### CODE REVIEW

File: lib/core/i18n/i18n.utils.ts [hunk 1]
```diff
@@ -9,7 +12,10 @@ import { Maybe } from "@trpc/server";
 import { i18n } from "../../../next-i18next.config";
 
 export function getLocaleFromHeaders(req: IncomingMessage): string {
-  const preferredLocale = parser.pick(i18n.locales, req.headers["accept-language"]) as Maybe<string>;
+  const preferredLocale = parser.pick(
+    i18n.locales,
+    req.headers["accept-language"] as string
+  ) as Maybe<string>;
```
Comment (zomars) [inline lib/core/i18n/i18n.utils.ts:18]: I think this will conflict with my PR: https://github.com/calendso/calendso/pull/988/files#r731431175

File: lib/core/i18n/i18n.utils.ts [hunk 2]
```diff
@@ -44,3 +50,18 @@ export const getOrSetUserLocaleFromHeaders = async (req: IncomingMessage): Promi
...
+    _i18n.init({
+      lng: locale,
+      resources: _nextI18Next.initialI18nStore,
+      fallbackLng: _nextI18Next.userConfig?.i18n.defaultLocale,
+    });
+    return _i18n;
+  };
+  const _i18n = nexti18next != null ? nexti18next : await create();
+  return _i18n.getFixedT(locale, ns);
+};
```
Comment (zomars) [inline lib/core/i18n/i18n.utils.ts:67]: What are we doing here?

Comment (mihaic195) [reply lib/core/i18n/i18n.utils.ts:67]: It's a way to make use of the `t` function on server-side code (e.g. api, emails, etc.).

File: lib/core/i18n/i18n.utils.ts (hunk 1 above)
Comment (mihaic195) [reply lib/core/i18n/i18n.utils.ts:18]: Not necessarily a problem, we can fix the conflict later depending which PR gets merged first.

**Review by pumfleet — APPROVED** [review]

File: lib/core/i18n/i18n.utils.ts [hunk 3]
```diff
@@ -44,3 +50,18 @@ export const getOrSetUserLocaleFromHeaders = async (req: IncomingMessage): Promi
 
   return preferredLocale;
 };
+
+export const getT = async (locale: string, ns: string) => {
```
Comment (KATT) [inline lib/core/i18n/i18n.utils.ts:54]: - More verbose naming please
- Move to `/server/*`-folder if it's server-only

File: lib/core/i18n/i18n.utils.ts (hunk 2 above)
Comment (KATT) [reply lib/core/i18n/i18n.utils.ts:67]: Cool solution, but isn't there not an easier way to get this? 🤔

**Review by KATT — COMMENTED** [review]
Looks decent - just rename that function to something more readable

File: lib/core/i18n/i18n.utils.ts (hunk 2 above)
Comment (emrysal) [reply lib/core/i18n/i18n.utils.ts:67]: Not sure there is @KATT, I've checked out a few blogs in this regard and it looks one of the "More complex" things to solve.

File: lib/emails/invitation.ts [hunk 4]
```diff
@@ -26,28 +28,30 @@ const sendEmail = (invitation: Invitation, provider: EmailProvider): Promise<voi
...
-          (invitation.from ? invitation.from + " invited you" : "You have been invited") +
-          ` to join ${invitation.teamName}`,
+        subject: invitation.from
+          ? t("user_invited_you", { user: invitation.from, teamName: invitation.teamName })
+          : t("you_have_been_invited", { teamName: invitation.teamName }),
         html: invitationHtml,
         text: text(invitationHtml),
       },
       (error) => {
         if (error) {
```
Comment (emrysal) [inline lib/emails/invitation.ts:44]: `if (error instanceof Error)` is better here.

File: lib/emails/invitation.ts [hunk 5]
```diff
@@ -26,28 +28,30 @@ const sendEmail = (invitation: Invitation, provider: EmailProvider): Promise<voi
...
           console.error("SEND_INVITATION_NOTIFICATION_ERROR", invitation.toEmail, error);
-          return reject(new Error(error));
+          return reject(new Error(error.message));
```
Comment (emrysal) [inline lib/emails/invitation.ts:46]: I'm actually relatively certain this previously handled the new Error(<string>error); as throwables can be an Error or string object - this is quite likely to fail (needs to handle the different types, string or Error) - also; this is also showing as a typescript error on my machine.

Comment (emrysal) [reply lib/emails/invitation.ts:46]: Passing an `Error` object to another new Error causes this: `Error: Error: <message>` - but calling .message on a string fails a bit harder, too.

**Review by emrysal — CHANGES_REQUESTED** [review]
Overall looks pretty good, got some concerns about a specific error handler so let's discuss.

File: lib/emails/invitation.ts (hunk 4 above)
Comment (zomars) [reply lib/emails/invitation.ts:44]: Does `if (_err) ` necessary? Can it be undefined or null?

File: pages/api/teams/[team]/invite.ts [hunk 6]
```diff
@@ -2,11 +2,13 @@ import { randomBytes } from "crypto";
 import type { NextApiRequest, NextApiResponse } from "next";
 
 import { getSession } from "@lib/auth";
-
-import { createInvitationEmail } from "../../../../lib/emails/invitation";
-import prisma from "../../../../lib/prisma";
+import { getT } from "@lib/core/i18n/i18n.utils";
```
Comment (zomars) [inline pages/api/teams/[team]/invite.ts:5]: Agreed with @KATT we could be more explicit with naming here

File: pages/auth/forgot-password/index.tsx [hunk 7]
```diff
@@ -157,7 +159,7 @@ export default function ForgotPassword({ csrfToken }) {
   );
 }
 
-ForgotPassword.getInitialProps = async (context) => {
+ForgotPassword.getInitialProps = async (context: GetServerSidePropsContext) => {
```
Comment (zomars) [inline pages/auth/forgot-password/index.tsx:162]: Isn't `getInitialProps` deprecated?

**Review by zomars — COMMENTED** [review]
Code looks good, can't think of more feedback apart from what @emrysal and @KATT have already addressed.

**Review by zomars — APPROVED** [review]
Nice work mate

**Review by KATT — COMMENTED** [review]
Unsure how to test but it looks like an improvement overall!

**Review by emrysal — APPROVED** [review]
Was assuming the typescript errors needed fixing first, but without those; this is good to go
